lighter/utils/freezer.py

```diff
@@ -0,0 +1,84 @@
+from typing import List, Optional, Union
+
+from loguru import logger
+from torch.nn import Module
+
+from lighter.utils.misc import ensure_list
+
+
+class LighterFreezer:
```

Comment:
I know we discussed this a bit before. But now that I think of it more, is there any reason this isn't a callback?

PL seems to have a very similar option https://github.com/Lightning-AI/lightning/blob/a028171f26d781491fa87fd27adf0e6dc1d7c3ae/pytorch_lightning/callbacks/finetuning.py#L38

Although it seems like some elements are not necessary and overcomplicate things. 

But I lean toward a callback if you think it makes sense, avoiding adding more logic into `system` for specific functionality

Reply:
We can apply the same approach to Inferers, but it's not the best solution. To keep things clear, let's include components related to the model's logic in LighterSystem and keep Callbacks like Loggers, Writers, and Checkpointers separate since they don't affect the model's logic. What do you think?

Reply:
While I agree with you, I think it will get more difficult as we scale. 

A few use cases I want to have in `lighter` as we move along.
- Auto-learning rate finder: https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.callbacks.LearningRateFinder.html
- Freeze and unfreeze based on losses. Doing this with the current freezer implementation will mean we need to pass in an additional object or self itself. Not really a problem, but why not just go to a callback that does this more systematically?
- Changing training parameters dynamically, like `temperature` in SSL. 
- Stochastic Weight Averaging, which I found to improve performance quite a bit for segmentation models 

These are all cases that affect the model logic. How do we effectively decide what will go into the system and what is added as a callback?
On that note, the inferer is quite specific as well, we mostly need it for medical imaging use cases. Nothing stops us from doing it for other-use cases, but it's uncommon. It might make more sense to have it as a callback in this case.

It's nice to have the users know where things are, but as we go along, we might have many such additions, and then it's more our personal choice, based on importance, that decides what we should expose to the user through the system and what is a callback. I find this confusing, why is the freezer more important to expose than an auto-learning rate finder?  These are both, in my opinion, additional features to a regular model training loop. 

I propose that we keep the system as simple as possible, only showing the core functionality of the model training loop. Everything else that modifies it (freezer, inferer, what I mentioned above)/supports it (checkpoints, logger, writers) goes as callbacks.

Reply:
Alright, I agree - let's make another PR for tha, I'll make the inferer and the freezer callbacks

Reply:
One thing that comes to my mind, what if you need callbacks in certain order? Do you think that could happen?

Reply:
https://github.com/Lightning-AI/lightning/issues/10260